http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java

```diff
@@ -0,0 +1,1263 @@
+                        emitter.error(customizeException(new HttpClientException("Connect Error: " + cause.getMessage(), cause)));
+                    } else {
+                        emitter.success(mockPoolHandle(connectionFuture.channel()));
+                    }
+                });
+            }
+        })
+            .delayUntil(this::delayUntilHttp2Ready)
+            .map(poolHandle -> {
+                addReadTimeoutHandler(poolHandle.channel.pipeline());
+                return poolHandle;
+            });
+    }
+
+    private Publisher<?> delayUntilHttp2Ready(PoolHandle poolHandle) {
+        Http2SettingsHandler settingsHandler = (Http2SettingsHandler) poolHandle.channel.pipeline().get(ChannelPipelineCustomizer.HANDLER_HTTP2_SETTINGS);
+        if (settingsHandler == null) {
+            return Flux.empty();
+        }
+        Sinks.Empty<?> empty = Sinks.empty();
+        addInstrumentedListener(settingsHandler.promise, future -> {
+            if (future.isSuccess()) {
+                empty.tryEmitEmpty();
+            } else {
+                poolHandle.taint();
+                poolHandle.release();
+                empty.tryEmitError(future.cause());
+            }
+        });
+        return empty.asMono();
+    }
+
+    Mono<PoolHandle> connectForStream(DefaultHttpClient.RequestKey requestKey, boolean isProxy, boolean acceptEvents) {
+        return Mono.<PoolHandle>create(emitter -> {
+            ChannelFuture channelFuture;
+            try {
+                if (httpVersion == HttpVersion.HTTP_2_0) {
+
+                    channelFuture = doConnect(requestKey, true, isProxy, acceptEvents, channelHandlerContext -> {
+                        try {
+                            final Channel channel = channelHandlerContext.channel();
+                            emitter.success(mockPoolHandle(channel));
+                        } catch (Exception e) {
+                            emitter.error(e);
+                        }
+                    });
+                } else {
+                    channelFuture = doConnect(requestKey, true, isProxy, acceptEvents, null);
+                    addInstrumentedListener(channelFuture,
+                        (ChannelFutureListener) f -> {
+                            if (f.isSuccess()) {
+                                Channel channel = f.channel();
+                                emitter.success(mockPoolHandle(channel));
+                            } else {
+                                Throwable cause = f.cause();
+                                emitter.error(customizeException(new HttpClientException("Connect error:" + cause.getMessage(), cause)));
+                            }
+                        });
+                }
+            } catch (HttpClientException e) {
+                emitter.error(e);
+                return;
+            }
+
+            // todo: on emitter dispose/cancel, close channel
```

Comment:
Is this TODO still to be done?

Reply:
I removed code from there that appeared to close the channel, but actually did nothing. I think the channel should be closed in this case, but I think that change would be too risky atm.